File: R/analysis.R
# ---------------------------------------------------------------------------
# Pipeline step 9: database-wide analysis.
#
# A deliberate, user-triggered escalation -- heavier than a single-document
# pass, and it does not wait for entity resolution. Matching is on normalised
# raw name text, which is a stepping stone to Phase 4's real resolution and not
# a substitute for it. Every result carries resolution_quality =
# "naive_unresolved" so nothing downstream can mistake it for resolved data.
#
# What that means in practice, and why it is stated on every result: the naive
# key matches "Ernst & Young LLP" to "Ernst & Young", and fails to match either
# to "Ernst and Young". It over-merges distinct entities that share a name and
# under-merges one entity written two ways. That is acceptable for a
# best-effort "has this name appeared elsewhere?" and unacceptable as a basis
# for a conflict-of-interest finding.
# ---------------------------------------------------------------------------

#' Quality label attached to every naive-matched result
#' @export
ORPH_NAIVE_RESOLUTION <- "naive_unresolved"

# ---------------------------------------------------------------------------
# Interface queries
# ---------------------------------------------------------------------------

#' Query every object type that implements an interface, as one set
#'
#' The ontology's job is to let a question be asked once. "Which named entities
#' appear in this document" spans Company and Person today and may span more
#' tomorrow; asking it per type means the list of types lives at the call site,
#' and a new type silently stops being included in the answer. Asking it by
#' interface means adding `"Named"` to a new object type is enough.
#'
#' Mirrors `objectSetsR::object_set_by_interface()`: the result is projected to
#' the interface's properties only, so every row has the same shape whichever
#' type it came from. A `type_id` column says which that was.
#'
#' @param con A connection.
#' @param interface_id Interface identifier, e.g. `"Named"`.
#' @param bundle A bundle. Defaults to the active one.
#' @param document_id Restrict to one document, or `NULL` for the corpus.
#' @param include_rejected Include rows a reviewer rejected.
#' @param where Optional extra SQL predicate over the interface's properties.
#' @param params Parameters for `where`.
#' @return A data frame with the interface's properties plus `type_id`.
#' @export
orph_object_set_by_interface <- function(con, interface_id, bundle = NULL,
                                         document_id = NULL, include_rejected = FALSE,
                                         where = NULL, params = list()) {
  bundle <- bundle %||% orph_active_bundle(con)
  if (is.null(bundle)) cli::cli_abort("No active ontology bundle.")

  iface <- orph_interface(bundle, interface_id)
  if (is.null(iface)) {
    known <- vapply(bundle$interfaces %||% list(), function(i) i$id %||% "", character(1))
    cli::cli_abort(c("No interface {.val {interface_id}} in the bundle.",
                     i = "Known interfaces: {.val {known}}"))
  }

  type_ids <- orph_implementing_types(bundle, interface_id)
  if (length(type_ids) == 0) {
    cli::cli_abort("No object type implements {.val {interface_id}}.")
  }

  cols <- orph_interface_property_ids(iface)
  selects <- character(); all_params <- list()

  for (tid in type_ids) {
    ot <- orph_object_type(bundle, tid)
    if (is.null(ot) || !DBI::dbExistsTable(con, ot$table_name)) next

    # Validated at registration, but a table can lag its bundle between an
    # amendment being accepted and the schema being applied.
    present <- DBI::dbListFields(con, ot$table_name)
    if (!all(cols %in% present)) next

    clauses <- character()
    if (!include_rejected) clauses <- c(clauses, "status != 'rejected'")
    if (!is.null(document_id)) { clauses <- c(clauses, "document_id = ?"); all_params <- c(all_params, document_id) }
    if (!is.null(where))       { clauses <- c(clauses, paste0("(", where, ")")); all_params <- c(all_params, params) }

    selects <- c(selects, sprintf(
      "SELECT %s, %s AS type_id FROM %s%s",
      paste(DBI::dbQuoteIdentifier(con, cols), collapse = ", "),
      DBI::dbQuoteString(con, tid),
      DBI::dbQuoteIdentifier(con, ot$table_name),
      if (length(clauses)) paste0(" WHERE ", paste(clauses, collapse = " AND ")) else ""))
  }

  if (length(selects) == 0) {
    empty <- stats::setNames(
      lapply(seq_along(c(cols, "type_id")), function(i) character()), c(cols, "type_id"))
    return(as.data.frame(empty, stringsAsFactors = FALSE))
  }

  db_query(con, paste(selects, collapse = "\nUNION ALL\n"), all_params)
}

#' Run the database-wide analysis for a document
#'
#' Answers the questions the architecture names for this step: do the companies
#' and people in this document appear in others, and how does this contract's
#' value sit against other contracts involving the same counterparties.
#'
#' @param con A writable connection.
#' @param document_id The document to compare against the rest of the corpus.
#' @param actor_id Actor triggering the escalation.
#' @param narrate Also produce a narrative reading of the findings. Requires a
#'   cloud opt-in when `tier = "cloud"`.
#' @param tier Tier for the optional narrative.
#' @param opt_in Explicit cloud opt-in for the narrative.
#' @return A list with the findings and the evaluation id they were stored under.
#' @export
orph_corpus_analysis <- function(con, document_id, actor_id = NULL, narrate = FALSE,
                                 tier = c("cloud", "local"), opt_in = FALSE) {
  assert_writable(con)
  tier   <- match.arg(tier)
  bundle <- orph_active_bundle(con)
  doc    <- orph_get_document(con, document_id)
  if (is.null(doc)) cli::cli_abort("No document {.val {document_id}}.")

  n_docs <- db_get_one(con, "SELECT COUNT(*) AS n FROM documents")$n
  if (n_docs < 2) {
    cli::cli_abort(c(
      "Database-wide analysis needs more than one document; the store has {n_docs}.",
      i = "Ingest and extract at least one more document first."))
  }

  findings <- list(
    counterparties = match_counterparties(con, bundle, document_id, "Company"),
    people         = match_counterparties(con, bundle, document_id, "Person")
  )
  findings$value_comparison <- compare_primary_values(con, bundle, document_id,
                                                      findings$counterparties)

  context_ids <- unique(c(
    unlist(lapply(findings$counterparties, function(f) f$other_instance_ids)),
    unlist(lapply(findings$people,         function(f) f$other_instance_ids))))
  local_ids <- unique(c(
    vapply(findings$counterparties, function(f) f$instance_id, character(1)),
    vapply(findings$people,         function(f) f$instance_id, character(1))))

  result <- list(
    document_id     = document_id,
    matched_companies = length(findings$counterparties),
    matched_people    = length(findings$people),
    counterparties  = findings$counterparties,
    people          = findings$people,
    value_comparison = findings$value_comparison,
    identifier_matched = sum(vapply(findings$counterparties,
                                    function(f) length(f$identifier_matches) > 0, logical(1))),
    caveat = paste(
      "Name matching is on normalised raw text, not resolved entities.",
      "Matches listed under identifier_matches rest on a stated registration",
      "number instead and are exact.",
      "Two different organisations sharing a name will be merged, and one",
      "organisation written two ways will not be. Treat these as leads to check,",
      "not as findings.")
  )

  narrative <- NULL
  if (isTRUE(narrate)) {
    if (tier == "cloud") orph_assert_cloud_allowed(con, opt_in = opt_in, actor_id = actor_id)
    reply <- orph_llm_json(con, tier, corpus_system_prompt(),
                           as.character(to_json(result)),
                           purpose = "corpus_analysis", document_id = document_id,
                           actor_id = actor_id, excerpt_only = TRUE, opt_in = opt_in)
    narrative <- list(summary = as_scalar_or_na(reply$summary),
                      observations = reply$observations %||% list(),
                      suggested_checks = reply$suggested_checks %||% list())
    result$narrative <- narrative
  }

  eval_id <- NULL
  with_tx(con, {
    eval_id <- write_concept_evaluation(con, bundle,
      concept_id = "corpus_comparison", concept_version = NA_integer_,
      concept_scope = NA_character_, kind = "corpus", scope_level = "database",
      document_id = document_id, result = result,
      dependencies = c(local_ids, context_ids),
      source = if (isTRUE(narrate) && tier == "cloud") "ai_cloud" else "ai_local",
      confidence = unname(ORPH_CONFIDENCE[["inferred"]]),
      actor_id = actor_id,
      corpus_context = list(instance_ids = context_ids,
                            document_ids = unique(unlist(lapply(
                              c(findings$counterparties, findings$people),
                              function(f) f$other_document_ids)))),
      resolution_quality = ORPH_NAIVE_RESOLUTION)
  })

  c(result, list(evaluation_id = eval_id, resolution_quality = ORPH_NAIVE_RESOLUTION))
}

#' @keywords internal
match_counterparties <- function(con, bundle, document_id, type_id) {
  # The instances in *this* document still come from one type -- the caller asks
  # about companies or about people. What changed is the lookup: it now searches
  # every type implementing `Named`, not just the same type. A name is a name,
  # and whether the match lands on a Company or a Person is a finding rather
  # than something to filter out in advance.
  ot <- orph_object_type(bundle, type_id)
  if (is.null(ot) || !DBI::dbExistsTable(con, ot$table_name)) return(list())

  mine <- db_query(con, sprintf(
    "SELECT instance_id, name, naive_key FROM %s
     WHERE document_id = ? AND status != 'rejected' AND naive_key IS NOT NULL AND naive_key != ''",
    DBI::dbQuoteIdentifier(con, ot$table_name)), list(document_id))
  if (nrow(mine) == 0) return(list())

  named <- orph_object_set_by_interface(con, "Named", bundle = bundle)
  if (nrow(named) == 0) return(list())
  named <- named[!is.na(named$naive_key) & nzchar(named$naive_key), , drop = FALSE]

  # A registration number is an identity claim; a normalised name is a guess.
  # Where both documents state one, matching on it is exact -- it neither
  # over-merges two companies that share a name nor under-merges one company
  # written two ways, which are the two failure modes the naive key has. It is
  # gathered separately and reported separately, because a match backed by an
  # identifier is much stronger evidence than a match backed by a name.
  registered <- identifier_matches(con, document_id)

  out <- list()
  for (i in seq_len(nrow(mine))) {
    key <- mine$naive_key[[i]]
    others <- named[named$naive_key %in% key & named$document_id != document_id, , drop = FALSE]
    by_identifier <- registered[[mine$instance_id[[i]]]] %||% list()

    # Either kind of match is worth reporting. Requiring a name match first
    # would have made the identifier useless: it exists precisely to catch the
    # entities normalisation misses -- "Meridian Systems Limited" and "Meridian
    # Sys. Ltd" share a registration number and nothing else.
    if (nrow(others) == 0 && length(by_identifier) == 0) next

    same_type  <- others[others$type_id == type_id, , drop = FALSE]
    other_type <- others[others$type_id != type_id, , drop = FALSE]
    variants   <- unique(c(others$name,
                           vapply(by_identifier, function(m) m$name, character(1))))

    out[[length(out) + 1L]] <- list(
      instance_id          = mine$instance_id[[i]],
      name                 = mine$name[[i]],
      naive_key            = key,
      appears_in_documents = length(unique(same_type$document_id)),
      other_document_ids   = unique(same_type$document_id),
      other_instance_ids   = unique(same_type$instance_id),
      name_variants        = variants,
      # Surfaced rather than hidden: differing spellings under one key are the
      # clearest signal that this needs real resolution.
      spelling_varies      = length(setdiff(variants, mine$name[[i]])) > 0,
      # Matched on a stated registration number rather than on a name. Where
      # this is non-empty the match does not rest on normalisation at all.
      identifier_matches   = by_identifier,
      # A name that is a company here and a person elsewhere is exactly the
      # shape of thing a reviewer wants to look at, so it is reported rather
      # than quietly dropped -- and reported separately, because it is a
      # weaker signal than a same-type match.
      cross_type_matches   = if (nrow(other_type) == 0) list() else lapply(
        seq_len(nrow(other_type)), function(k) list(
          type_id     = other_type$type_id[[k]],
          name        = other_type$name[[k]],
          instance_id = other_type$instance_id[[k]],
          document_id = other_type$document_id[[k]]))
    )
  }
  out
}

#' Match this document's companies to others by stated registration number
#'
#' Exact rather than heuristic. Returns a list keyed by this document's
#' instance id, so a caller can attach the matches to the entity they belong to.
#'
#' @keywords internal
identifier_matches <- function(con, document_id) {
  if (!DBI::dbExistsTable(con, "instances_Company")) return(list())
  cols <- DBI::dbListFields(con, "instances_Company")
  if (!("registration_number" %in% cols)) return(list())

  mine <- db_query(con,
    "SELECT instance_id, name, registration_number FROM instances_Company
     WHERE document_id = ? AND status != 'rejected'
       AND registration_number IS NOT NULL AND registration_number != ''",
    list(document_id))
  if (nrow(mine) == 0) return(list())

  out <- list()
  for (i in seq_len(nrow(mine))) {
    others <- db_query(con,
      "SELECT instance_id, document_id, name, registration_number
       FROM instances_Company
       WHERE registration_number = ? AND document_id != ? AND status != 'rejected'",
      list(mine$registration_number[[i]], document_id))
    if (nrow(others) == 0) next
    out[[mine$instance_id[[i]]]] <- lapply(seq_len(nrow(others)), function(k) list(
      instance_id         = others$instance_id[[k]],
      document_id         = others$document_id[[k]],
      name                = others$name[[k]],
      registration_number = others$registration_number[[k]],
      # Recorded because it is the interesting case: the same registered entity
      # written two different ways is precisely what entity resolution is for,
      # and here it is proven rather than guessed.
      name_differs        = !identical(others$name[[k]], mine$name[[i]])))
  }
  out
}

#' Compare this document's headline value against related documents
#'
#' The type and the properties come from the bundle's domain block. A domain
#' with no comparable magnitude declares none, and this reports itself
#' unavailable rather than being absent.
#'
#' @keywords internal
compare_primary_values <- function(con, bundle, document_id, counterparties) {
  domain <- orph_domain(bundle)
  if (is.null(domain$primary_object_type) || is.null(domain$value_property)) {
    return(list(available = FALSE,
                reason = "This bundle declares no comparable value for its primary object type."))
  }
  primary <- orph_object_type(bundle, domain$primary_object_type)
  if (is.null(primary) || !DBI::dbExistsTable(con, primary$table_name)) {
    return(list(available = FALSE, reason = "No instances of the primary object type exist."))
  }

  tbl <- DBI::dbQuoteIdentifier(con, primary$table_name)
  val <- DBI::dbQuoteIdentifier(con, domain$value_property)
  has_ccy <- !is.null(domain$currency_property)
  ccy <- if (has_ccy) DBI::dbQuoteIdentifier(con, domain$currency_property) else NULL

  select_cols <- paste(c("instance_id", "document_id", as.character(val),
                         if (has_ccy) as.character(ccy)), collapse = ", ")

  this <- db_get_one(con, sprintf(
    "SELECT %s FROM %s WHERE document_id = ? AND status != 'rejected'
     ORDER BY confidence DESC LIMIT 1", select_cols, tbl), list(document_id))
  if (is.null(this) || is.na(this[[domain$value_property]])) {
    return(list(available = FALSE,
                reason = sprintf("No %s has been extracted for this document.",
                                 domain$value_property)))
  }

  other_docs <- unique(unlist(lapply(counterparties, function(f) f$other_document_ids)))
  if (length(other_docs) == 0) {
    return(list(available = FALSE,
                reason = "No other documents share a counterparty name with this one."))
  }

  peers <- db_query(con, sprintf(
    "SELECT %s FROM %s
     WHERE status != 'rejected' AND %s IS NOT NULL AND document_id IN (%s)",
    select_cols, tbl, val, paste(rep("?", length(other_docs)), collapse = ", ")),
    as.list(other_docs))
  if (nrow(peers) == 0) {
    return(list(available = FALSE,
                reason = "Related documents have no extracted value to compare."))
  }

  mixed <- FALSE
  if (has_ccy) {
    # Values are compared only within a currency: converting them would need a
    # rate for the right date, which is not something to invent here.
    same <- peers[!is.na(peers[[domain$currency_property]]) &
                  peers[[domain$currency_property]] ==
                    (this[[domain$currency_property]] %||% ""), , drop = FALSE]
    mixed <- nrow(same) < nrow(peers)
    peers <- same
  }

  values <- suppressWarnings(as.numeric(peers[[domain$value_property]]))
  values <- values[!is.na(values)]
  mine   <- suppressWarnings(as.numeric(this[[domain$value_property]]))
  if (length(values) == 0) {
    return(list(available = FALSE,
                reason = "Related documents are in other currencies; no conversion is applied."))
  }

  list(
    available       = TRUE,
    object_type     = domain$primary_object_type,
    value_property  = domain$value_property,
    this_value      = mine,
    currency        = if (has_ccy) this[[domain$currency_property]] else NA_character_,
    peer_count      = length(values),
    peer_median     = stats::median(values),
    peer_max        = max(values),
    peer_min        = min(values),
    ratio_to_median = if (stats::median(values) > 0) round(mine / stats::median(values), 2) else NA,
    mixed_currencies_excluded = mixed
  )
}

#' @keywords internal
corpus_system_prompt <- function() {
  paste0(
    "You are advising a public servant comparing one contract against others in ",
    "their corpus. You are given best-effort matches made by normalising raw name ",
    "text -- not resolved entities.\n\n",
    "Return JSON:\n",
    '{"summary":"...","observations":["..."],"suggested_checks":["..."]}\n\n',
    "Rules:\n",
    "- Treat every match as provisional. Say 'a company with this name' rather than asserting identity.\n",
    "- Where name spellings differ under one match, call that out as needing confirmation.\n",
    "- Do not assert a conflict of interest. Suggest what a person should check."
  )
}

File: orpheus/analysis.py
"""Step 9: what this document looks like against the rest of the corpus.

Deliberately labelled as a stepping stone. Matching is on **normalised raw
text**, not resolved entities, and every result carries `naive_unresolved` so
nothing downstream can mistake it for entity resolution. Two organisations
sharing a name will be merged; one organisation written two ways will not be.
Those are leads to check, not findings.

The one exact path is a stated registration number. A number is an identity
claim where a normalised name is a guess, so identifier matches are gathered and
reported **separately** — a match resting on an identifier is much stronger
evidence than one resting on spelling, and collapsing the two would hide that.

Nothing here knows what a contract is. Which type a document is about, and which
property holds a comparable magnitude, come from the bundle's domain block; a
domain with no comparable value says so rather than being silently absent.
"""

from __future__ import annotations

import statistics

from . import bundle as bundle_mod
from .concepts import write_evaluation
from .ingest import get_document
from .rubric import CONFIDENCE, NAIVE_RESOLUTION
from .store import Store
from .utils import NotFound, OrpheusError, to_json


def object_set_by_interface(store: Store, interface_id: str,
                            bundle: dict | None = None,
                            document_id: str | None = None,
                            include_rejected: bool = False,
                            where: str | None = None,
                            params: tuple = ()) -> list[dict]:
    """Every type implementing an interface, as one set.

    An interface is a promise that several types can answer the same question.
    This is where the promise gets used: a name is a name, whether it belongs to
    a Company or a Person, and asking "what else in the corpus is called this?"
    should not have to know which.
    """
    bundle = bundle or bundle_mod.active(store) or bundle_mod.load()
    interface = bundle_mod.interface(bundle, interface_id)
    if interface is None:
        known = ", ".join(i["id"] for i in bundle.get("interfaces", []))
        raise NotFound(
            f"No interface {interface_id!r} in the bundle. Known: {known}.")

    type_ids = bundle_mod.implementing_types(bundle, interface_id)
    if not type_ids:
        raise OrpheusError(f"No object type implements {interface_id!r}.")

    columns = bundle_mod.interface_property_ids(interface)
    selects: list[str] = []
    all_params: list = []

    for type_id in type_ids:
        obj = bundle_mod.object_type(bundle, type_id)
        table = bundle_mod.table_name(obj) if obj else None
        if not table or not store.table_exists(table):
            continue
        # Validated at registration, but a table can lag its bundle between an
        # amendment being accepted and the schema being applied.
        if not set(columns) <= set(store.columns(table)):
            continue

        clauses = []
        if not include_rejected:
            clauses.append("status != 'rejected'")
        if document_id is not None:
            clauses.append("document_id = ?")
            all_params.append(document_id)
        if where:
            clauses.append(f"({where})")
            all_params.extend(params)

        projection = ", ".join(f'"{c}"' for c in columns)
        selects.append(
            f"SELECT {projection}, '{type_id}' AS type_id FROM \"{table}\""
            + (" WHERE " + " AND ".join(clauses) if clauses else ""))

    if not selects:
        return []
    return store.query("\nUNION ALL\n".join(selects), tuple(all_params))


def identifier_matches(store: Store, document_id: str) -> dict[str, list[dict]]:
    """Companies matched to other documents by stated registration number.

    Exact rather than heuristic, and gathered separately so a caller can attach
    each match to the entity it belongs to.
    """
    if not store.table_exists("instances_Company"):
        return {}
    if "registration_number" not in store.columns("instances_Company"):
        return {}

    mine = store.query(
        "SELECT instance_id, name, registration_number FROM instances_Company "
        "WHERE document_id = ? AND status != 'rejected' "
        "AND registration_number IS NOT NULL AND registration_number != ''",
        (document_id,))

    out: dict[str, list[dict]] = {}
    for row in mine:
        others = store.query(
            "SELECT instance_id, document_id, name, registration_number "
            "FROM instances_Company WHERE registration_number = ? "
            "AND document_id != ? AND status != 'rejected'",
            (row["registration_number"], document_id))
        if not others:
            continue
        out[row["instance_id"]] = [{
            "instance_id": other["instance_id"],
            "document_id": other["document_id"],
            "name": other["name"],
            "registration_number": other["registration_number"],
            # The interesting case: the same registered entity written two
            # different ways is precisely what entity resolution is for, and
            # here it is proven rather than guessed.
            "name_differs": other["name"] != row["name"],
        } for other in others]
    return out


def match_counterparties(store: Store, bundle: dict, document_id: str,
                         type_id: str) -> list[dict]:
    obj = bundle_mod.object_type(bundle, type_id)
    table = bundle_mod.table_name(obj) if obj else None
    if not table or not store.table_exists(table):
        return []

    mine = store.query(
        f'SELECT instance_id, name, naive_key FROM "{table}" '
        "WHERE document_id = ? AND status != 'rejected' "
        "AND naive_key IS NOT NULL AND naive_key != ''", (document_id,))
    if not mine:
        return []

    # Searches every type implementing Named, not just the same type: whether a
    # match lands on a Company or a Person is a finding, not something to filter
    # out in advance.
    named = [row for row in object_set_by_interface(store, "Named", bundle=bundle)
             if row.get("naive_key")]
    registered = identifier_matches(store, document_id)

    out = []
    for row in mine:
        # An entity link, where one exists, is what this question is *for*: a
        # person decided these mentions are the same thing, which is strictly
        # better evidence than the normalised name that was only ever a
        # stand-in for the decision. Re-deriving it from `naive_key` here would
        # ignore the answer and recompute a worse one.
        linked = entity_siblings(store, row["instance_id"], document_id)
        others = linked or [n for n in named
                            if n["naive_key"] == row["naive_key"]
                            and n["document_id"] != document_id]
        by_identifier = registered.get(row["instance_id"], [])
        # Either kind of match is worth reporting. Requiring a name match first
        # would make the identifier useless — it exists precisely to catch the
        # entities normalisation misses.
        if not others and not by_identifier:
            continue

        same_type = [o for o in others if o["type_id"] == type_id]
        other_type = [o for o in others if o["type_id"] != type_id]
        variants = list(dict.fromkeys(
            [o["name"] for o in others] + [m["name"] for m in by_identifier]))

        out.append({
            "instance_id": row["instance_id"],
            "name": row["name"],
            "naive_key": row["naive_key"],
            "appears_in_documents": len({o["document_id"] for o in same_type}),
            "other_document_ids": list(dict.fromkeys(o["document_id"] for o in same_type)),
            "other_instance_ids": list(dict.fromkeys(o["instance_id"] for o in same_type)),
            "name_variants": variants,
            # Surfaced rather than hidden: differing spellings under one key are
            # the clearest signal that this needs real resolution.
            "spelling_varies": bool(set(variants) - {row["name"]}),
            "identifier_matches": by_identifier,
            # Which path produced this match. A reviewed entity link and a
            # normalised name are not the same claim, and a caller that treats
            # them alike loses the only distinction that matters here.
            "matched_via": "entity" if linked else "naive_key",
            # A name that is a company here and a person elsewhere is exactly
            # what a reviewer wants to look at — reported, and reported
            # separately, because it is weaker than a same-type match.
            "cross_type_matches": [
                {"type_id": o["type_id"], "name": o["name"],
                 "instance_id": o["instance_id"], "document_id": o["document_id"]}
                for o in other_type],
        })
    return out


def entity_siblings(store: Store, instance_id: str, document_id: str) -> list[dict]:
    """Other mentions of the same entity, if this mention has been linked.

    Returns nothing when it has not, so the caller falls back to name matching.
    Only links a person has not rejected count: an unlinked or rejected one is
    a decision that these are *not* the same thing, and honouring it is the
    point of having asked.
    """
    if not store.table_exists("entity_mentions"):
        return []
    home = store.one(
        "SELECT entity_id FROM entity_mentions WHERE instance_id = ? "
        "AND unlinked_at IS NULL", (instance_id,))
    if home is None:
        return []

    out = []
    for row in store.query(
            "SELECT m.instance_id, m.document_id, i.type_id, i.table_name "
            "FROM entity_mentions m JOIN instance_index i USING (instance_id) "
            "WHERE m.entity_id = ? AND m.unlinked_at IS NULL "
            "AND m.document_id != ?", (home["entity_id"], document_id)):
        detail = store.one(
            f'SELECT name, naive_key FROM "{row["table_name"]}" '
            "WHERE instance_id = ?", (row["instance_id"],))
        out.append({"instance_id": row["instance_id"],
                    "document_id": row["document_id"],
                    "type_id": row["type_id"],
                    "name": detail["name"] if detail else None,
                    "naive_key": detail["naive_key"] if detail else None})
    return out


def compare_primary_values(store: Store, bundle: dict, document_id: str,
                           counterparties: list[dict]) -> dict:
    """This document's headline value against related documents.

    Every "unavailable" here is a distinct, stated reason rather than an empty
    result, because "no comparable value in this domain" and "no peer documents
    to compare against" call for different actions.
    """
    domain = bundle_mod.domain(bundle)
    value_property = domain.get("valueProperty")
    if not domain.get("primaryObjectType") or not value_property:
        return {"available": False,
                "reason": "This bundle declares no comparable value for its "
                          "primary object type."}

    primary = bundle_mod.object_type(bundle, domain["primaryObjectType"])
    table = bundle_mod.table_name(primary) if primary else None
    if not table or not store.table_exists(table):
        return {"available": False,
                "reason": "No instances of the primary object type exist."}

    currency_property = domain.get("currencyProperty")
    columns = ["instance_id", "document_id", value_property]
    if currency_property:
        columns.append(currency_property)
    projection = ", ".join(f'"{c}"' for c in columns)

    this = store.one(
        f'SELECT {projection} FROM "{table}" WHERE document_id = ? '
        "AND status != 'rejected' ORDER BY confidence DESC LIMIT 1", (document_id,))
    if this is None or this.get(value_property) is None:
        return {"available": False,
                "reason": f"No {value_property} has been extracted for this document."}

    peer_documents = list(dict.fromkeys(
        d for f in counterparties for d in f["other_document_ids"]))
    if not peer_documents:
        return {"available": False,
                "reason": "No other documents share a counterparty name with this one."}

    placeholders = ", ".join("?" for _ in peer_documents)
    peers = store.query(
        f'SELECT {projection} FROM "{table}" WHERE status != \'rejected\' '
        f'AND "{value_property}" IS NOT NULL AND document_id IN ({placeholders})',
        tuple(peer_documents))
    if not peers:
        return {"available": False,
                "reason": "Related documents have no extracted value to compare."}

    mixed = False
    if currency_property:
        # Compared only within a currency. Converting would need a rate for the
        # right date, which is not something to invent here.
        same = [p for p in peers
                if p.get(currency_property) == this.get(currency_property)]
        mixed = len(same) < len(peers)
        peers = same

    values = []
    for peer in peers:
        try:
            values.append(float(peer[value_property]))
        except (TypeError, ValueError):
            continue
    if not values:
        return {"available": False,
                "reason": "Related documents are in other currencies; no "
                          "conversion is applied."}

    try:
        mine = float(this[value_property])
    except (TypeError, ValueError):
        return {"available": False,
                "reason": f"This document's {value_property} is not numeric."}

    median = statistics.median(values)
    return {
        "available": True,
        "object_type": domain["primaryObjectType"],
        "value_property": value_property,
        "this_value": mine,
        "currency": this.get(currency_property) if currency_property else None,
        "peer_count": len(values),
        "peer_median": median,
        "peer_max": max(values),
        "peer_min": min(values),
        "ratio_to_median": round(mine / median, 2) if median > 0 else None,
        "mixed_currencies_excluded": mixed,
    }


CAVEAT = (
    "Name matching is on normalised raw text, not resolved entities. Matches "
    "listed under identifier_matches rest on a stated registration number "
    "instead and are exact. Two different organisations sharing a name will be "
    "merged, and one organisation written two ways will not be. Treat these as "
    "leads to check, not as findings."
)


def _resolution_quality(findings: list[dict]) -> str:
    if findings and all(f["matched_via"] == "entity" for f in findings):
        return "resolved"
    return NAIVE_RESOLUTION


def corpus_analysis(store: Store, document_id: str, actor_id: str | None = None,
                    narrate: bool = False, tier: str = "cloud",
                    opt_in: bool = False) -> dict:
    """Compare one document against the rest of the store."""
    store.assert_writable()
    bundle = bundle_mod.active(store) or bundle_mod.load()
    if get_document(store, document_id) is None:
        raise NotFound(f"No document {document_id!r}.")

    n_documents = store.scalar("SELECT COUNT(*) FROM documents") or 0
    if n_documents < 2:
        raise OrpheusError(
            f"Database-wide analysis needs more than one document; the store "
            f"has {n_documents}. Ingest and extract at least one more first.")

    counterparties = match_counterparties(store, bundle, document_id, "Company")
    people = match_counterparties(store, bundle, document_id, "Person")
    value_comparison = compare_primary_values(store, bundle, document_id,
                                              counterparties)

    result = {
        "document_id": document_id,
        "matched_companies": len(counterparties),
        "matched_people": len(people),
        "counterparties": counterparties,
        "people": people,
        "value_comparison": value_comparison,
        "identifier_matched": sum(1 for f in counterparties if f["identifier_matches"]),
        # `naive_unresolved` unless *every* match came through a reviewed entity
        # link. One name-derived match is enough to make the whole result a
        # stepping stone again, and saying otherwise would let a caller trust
        # the weakest row in the set.
        "resolution_quality": _resolution_quality(counterparties + people),
        "caveat": CAVEAT if any(f["matched_via"] == "naive_key"
                                for f in counterparties + people) else None,
    }

    if narrate:
        result["narrative"] = _narrate(store, result, document_id, actor_id,
                                       tier, opt_in)

    local_ids = [f["instance_id"] for f in counterparties + people]
    context_ids = [i for f in counterparties + people for i in f["other_instance_ids"]]

    with store.transaction():
        evaluation_id = write_evaluation(
            store, concept_id="corpus_comparison", concept_version=None,
            concept_scope=None, kind="corpus", scope_level="database",
            document_id=document_id, result=result,
            dependencies=local_ids + context_ids,
            source="ai_cloud" if (narrate and tier == "cloud") else "ai_local",
            confidence=CONFIDENCE["inferred"], actor_id=actor_id,
            corpus_context={"instance_ids": context_ids,
                            "document_ids": list(dict.fromkeys(
                                d for f in counterparties + people
                                for d in f["other_document_ids"]))},
            resolution_quality=result["resolution_quality"])

    result["evaluation_id"] = evaluation_id
    return result


NARRATIVE_PROMPT = (
    "You are given the result of a naive cross-document comparison. Summarise "
    "what a reviewer should look at. Return JSON only with: summary (one "
    "paragraph), observations (a list of short strings), suggested_checks (a "
    "list of short strings).\n"
    "The name matching is not entity resolution. Do not assert that two "
    "entities are the same organisation; say what would need checking to "
    "establish it."
)


def _narrate(store: Store, result: dict, document_id: str,
             actor_id: str | None, tier: str, opt_in: bool) -> dict:
    from .classify import _ask

    reply = _ask(store, tier, NARRATIVE_PROMPT, to_json(result) or "{}",
                 document_id=document_id, actor_id=actor_id,
                 excerpt_only=True, opt_in=opt_in)
    return {"summary": reply.get("summary"),
            "observations": reply.get("observations") or [],
            "suggested_checks": reply.get("suggested_checks") or []}
